### PR TITLE
Add support for sha256 field in matchspec

### LINKED ISSUE #12654: Allow sha256 to be included in match spec
### Checklist

- [X] I added a descriptive title
- [X] I searched open requests and couldn't find a duplicate

### What is the idea?

Currently a package match specification can include the md5 of the package to fully specify the package, for example `python[md5=bd5ef8b4531bcd58ac2c00393a7d8c3f]`.  It would be nice if the sha256 hash could be supplied instead.

### Why is this needed?

This would be beneficial as md5 is no longer considered a safe hash function and replacing it with another hash is often recommended. 

### What should happen?

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by jaimergp**
### Description

Closes #12654.

This was simpler than expected!

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [x] Add / update outdated documentation?

### CODE REVIEW

**Review by beeankha — DISMISSED** [review]

**Comment by jaimergp** [comment]
Added a snippet to [autodocument](https://docs.conda.io/projects/conda/en/latest/user-guide/concepts/pkg-search.html) the accepted fields (and their validators). Works locally and mimics what we do for the Configuration page.

This is how it looks:

image


(`sphinx.programoutput` assumes you want a code block, no way around that :/)

File: docs/source/user-guide/concepts/pkg-search.rst [hunk 1]
```diff
@@ -40,6 +40,11 @@ example image. The search below will return the same list of packages as the sta
 
    $ conda search "numpy[channel=conda-forge, subdir=linux-64, version=1.17.*, build=py38*]"
 
+This notation supports the following key-value pairs:
+
+.. program-output:: cd .. && python -c 'import os; import sys; src_dir = here = os.path.abspath(os.path.dirname("../setup.py")); sys.path.insert(0, src_dir); import conda.models.match_spec as M; print("-", "\n- ".join([f"{field:20} # validated via {M._implementors.get(field, M.ExactStrMatch).__name__}" for field in sorted(M.MatchSpec.FIELD_NAMES)]))'
```
Comment (jezdez) [inline docs/source/user-guide/concepts/pkg-search.rst:45]: Okay, `setup.py` doesn't exist anymore, should we use `pyproject.toml` to locate the conda folder instead? Why is this needed in the first place?

File: tests/models/test_match_spec.py [hunk 2]
```diff
@@ -1216,32 +1216,37 @@ def test_build_number_merge(self):
...
+            specs = (
+                MatchSpec(f"python[{hash_type}=FFBADD11]"),
+                MatchSpec("python=1.2.3"),
+                MatchSpec(f"python[{hash_type}=ffbadd11]"),
+            )
+            with pytest.raises(ValueError):
+                MatchSpec.merge(specs)
+
+    def test_hash_merge_wo_name(self):
+        for hash_type in ("md5", "sha256"):
```
Comment (jezdez) [inline tests/models/test_match_spec.py:1239]: This would be an ideal case for a parameterized pytest test, but may be out of scope.

OTOH I think we could port all tests in this module to pytest? /cc @kenodegard WDYT?

Comment (jaimergp) [reply tests/models/test_match_spec.py:1239]: Yea I tried that first but methods are not parameterizeable directly :/

File: docs/source/user-guide/concepts/pkg-search.rst (hunk 1 above)
Comment (jaimergp) [reply docs/source/user-guide/concepts/pkg-search.rst:45]: I just copied things from the configuration.rst page. I don't even think `setup.py` is being used at all there; maybe it was just a documentation thing (so we know it's the folder containing setup.py?). `src_dir = os.path.abspath("..")` should be enough.

And I guess the whole thing is to have development `conda` in PYTHONPATH if it's not installed, but maybe this is not needed anymore in the current workflow?

Comment (jezdez) [reply docs/source/user-guide/concepts/pkg-search.rst:45]: I think it's not needed anymore since we've modernized the documentation building last year, but I could be mistaken. Does it still work if you remove it?

File: tests/models/test_match_spec.py (hunk 2 above)
Comment (jezdez) [reply tests/models/test_match_spec.py:1239]: If you feel adventurous I think it should be easy to convert to function-based tests.

**Review by jezdez — COMMENTED** [review]
LGTM, with comments

File: tests/models/test_match_spec.py (hunk 2 above)
Comment (jaimergp) [reply tests/models/test_match_spec.py:1239]: Let's do that in a follow-up PR to keep this changeset cleaner. I can see how other implementations might point to this one for reference.

File: docs/source/user-guide/concepts/pkg-search.rst (hunk 1 above)
Comment (jaimergp) [reply docs/source/user-guide/concepts/pkg-search.rst:45]: It works as long as `conda` is installed in the docs environment, in editable mode. Otherwise you get this error, silently:

image

Comment (jaimergp) [reply docs/source/user-guide/concepts/pkg-search.rst:45]: I guess ReadTheDocs has been configured to install the project in the build env, so I'll remove it.

**Review by jezdez — APPROVED** [review]

**Comment by jezdez** [comment]
The error is a legitimate problem in how the signature verification tests are executed if conda-content-trust is broken: https://github.com/conda/conda/issues/12755

File: docs/source/user-guide/concepts/pkg-search.rst (hunk 1 above)
Comment (jezdez) [reply docs/source/user-guide/concepts/pkg-search.rst]: FWIW there was a regression about this after all: #12768
